src/client/cli/cmd/restore.cpp

```diff
@@ -0,0 +1,150 @@
+                                         "automatically discard current machine state.");
+
+            client->Write(client_response);
+            spinner.start();
+        }
+    };
+
+    return dispatch(&RpcMethod::restore, request, on_success, on_failure, streaming_callback);
+}
+
+std::string cmd::Restore::name() const
+{
+    return "restore";
+}
+
+QString cmd::Restore::short_help() const
+{
+    return QStringLiteral("Restore an instance from a snapshot");
+}
+
+QString cmd::Restore::description() const
+{
+    return QStringLiteral("Restore an instance to the state of a previously taken snapshot.");
+}
+
+mp::ParseCode cmd::Restore::parse_args(mp::ArgParser* parser)
+{
+    parser->addPositionalArgument("instance.snapshot",
+                                  "The instance to restore and snapshot to use, in <instance>.<snapshot> format, where "
+                                  "<instance> is the name of an instance, and <snapshot> is the name of a snapshot",
+                                  "<instance>.<snapshot>");
+
+    QCommandLineOption destructive({"d", "destructive"}, "Discard the current state of the instance");
+    parser->addOption(destructive);
+
+    auto status = parser->commandParse(this);
+    if (status != ParseCode::Ok)
+        return status;
+
+    const auto positional_args = parser->positionalArguments();
+    const auto num_args = positional_args.count();
+    if (num_args < 1)
+    {
+        cerr << "Need the name of an instance and snapshot to restore.\n";
+        return ParseCode::CommandLineError;
+    }
+
+    if (num_args > 1)
+    {
+        cerr << "Too many arguments supplied.\n";
+        return ParseCode::CommandLineError;
+    }
+
+    const auto tokens = parser->positionalArguments().at(0).split('.');
+    if (tokens.size() != 2 || tokens[0].isEmpty() || tokens[1].isEmpty())
+    {
+        cerr << "Invalid format. Please specify the instance to restore and snapshot to use in the form "
+                "<instance>.<snapshot>.\n";
+        return ParseCode::CommandLineError;
+    }
+
+    request.set_instance(tokens[0].toStdString());
+    request.set_snapshot(tokens[1].toStdString());
+    request.set_destructive(parser->isSet(destructive));
+    request.set_verbosity_level(parser->verbosityLevel());
+
+    return ParseCode::Ok;
+}
+
+bool cmd::Restore::confirm_destruction(const std::string& instance_name)
+{
+    static constexpr auto prompt_text =
+        "Do you want to take a snapshot of {} before discarding its current state? (Yes/no)";
+    static constexpr auto invalid_input = "Please answer Yes/no";
+    mp::PlainPrompter prompter(term);
+
+    auto answer = prompter.prompt(fmt::format(prompt_text, instance_name));
+    while (!answer.empty() && !std::regex_match(answer, yes_answer) && !std::regex_match(answer, no_answer))
+        answer = prompter.prompt(invalid_input);
+
+    return std::regex_match(answer, no_answer);
+}
```

Comment:
Nothing needs to be done at this moment, but this is quite similar to the code in `launch` for it's prompt around asking for permission to create the bridge.  I know @luis4a0 had done some refactoring in https://github.com/canonical/multipass/pull/3195, so we'll just need to consider how we may be able to refactor some of this once this PR is merged.